pipeline-model-definition/src/main/resources/org/jenkinsci/plugins/pipeline/modeldefinition/ModelInterpreter.groovy

```diff
@@ -209,22 +209,38 @@ class ModelInterpreter implements Serializable {
                                 script.parallel(getParallelStages(root, parentAgent, thisStage, firstError, parentStage, false, false, true))
                             }
                         } else {
-                            inDeclarativeAgent(thisStage, root, thisStage.agent) {
-                                if (evaluateWhen(thisStage.when)) {
-                                    withCredentialsBlock(thisStage.environment) {
-                                        withEnvBlock(thisStage.getEnvVars(script)) {
-                                            toolsBlock(thisStage.agent ?: root.agent, thisStage.tools, root) {
-                                                // Execute the actual stage and potential post-stage actions
-                                                executeSingleStage(root, thisStage, parentAgent)
-                                            }
+                            def stageBody = {
+                                withCredentialsBlock(thisStage.environment) {
+                                    withEnvBlock(thisStage.getEnvVars(script)) {
+                                        toolsBlock(thisStage.agent ?: root.agent, thisStage.tools, root) {
+                                            // Execute the actual stage and potential post-stage actions
+                                            executeSingleStage(root, thisStage, parentAgent)
                                         }
                                     }
-                                } else {
-                                    Utils.logToTaskListener("Stage '${thisStage.name}' skipped due to when conditional")
-                                    Utils.markStageSkippedForConditional(thisStage.name)
-                                    isSkipped = true
                                 }
                             }
+                            // If beforeAgent is true, evaluate the when before entering the agent.
+                            boolean whenPassed = false
+                            if (thisStage.when?.beforeAgent != null && thisStage.when?.beforeAgent) {
```

Comment:
I think this can be shortened to `thisStage.when?.beforeAgent` as groovy evaluates `null` as `false`.